channeldb/kvdb/etcd/commit_queue.go

```diff
@@ -0,0 +1,151 @@
+// +build kvdb_etcd
+
+package etcd
+
+import (
+	"context"
+	"sync"
+)
+
+// commitQueueSize is the maximum number of commits we let to queue up. All
+// remaining commits will block on commitQueue.Add().
+const commitQueueSize = 100
+
+// commitQueue is a simple execution queue to manage conflicts for transactions
+// and thereby reduce the number of times conflicting transactions need to be
+// retried. When a new transaction is added to the queue, we first upgrade the
+// read/write counts in the queue's own accounting to decide whether the new
+// transaction has any conflicting dependencies. If the transaction does not
+// conflict with any other, then it is comitted immediately, otherwise it'll be
+// queued up for later exection.
+// The algorithm is described in: http://www.cs.umd.edu/~abadi/papers/vll-vldb13.pdf
+type commitQueue struct {
+	ctx       context.Context
+	mx        sync.Mutex
+	readerMap map[string]int
+	writerMap map[string]int
+
+	commitMutex sync.RWMutex
+	queue       chan (func())
+	wg          sync.WaitGroup
+}
+
+// NewCommitQueue creates a new commit queue, with the passed abort context.
+func NewCommitQueue(ctx context.Context) *commitQueue {
+	q := &commitQueue{
+		ctx:       ctx,
+		readerMap: make(map[string]int),
+		writerMap: make(map[string]int),
+		queue:     make(chan func(), commitQueueSize),
+	}
+
+	// Start the queue consumer loop.
+	q.wg.Add(1)
+	go q.mainLoop()
+
+	return q
+}
+
+// Wait waits for the queue to stop (after the queue context has been canceled).
+func (c *commitQueue) Wait() {
+	c.wg.Wait()
+}
+
+// Add increases lock counts and queues up tx commit closure for execution.
+// Transactions that don't have any conflicts are executed immediately by
+// "downgrading" the count mutex to allow concurrency.
+func (c *commitQueue) Add(commitLoop func(), rset readSet, wset writeSet) {
+	c.mx.Lock()
+	blocked := false
+
+	// Increase read counts and mark as blocked if any key is in the
+	// writer map.
+	for key := range rset {
+		c.readerMap[key] += 1
+		blocked = blocked || c.writerMap[key] > 0
+	}
+
+	// Increase write counts and mark as blocked if any key is in the
+	// reader map or is already blocked for writing.
+	for key := range wset {
+		if !blocked {
+			// Check the read set because if we this transaction
+			// is the only one reading the key then we may not need
+			// to queue it up.
+			_, keyRead := rset[key]
+
+			// Transaction is blocked if:
+			// - there's any reader (which is not this tx).
+			// - there's any writer.
+			blocked = blocked || (c.readerMap[key] > 0 && !keyRead)
```

Comment:
still not sure if this is correct. Say this tx reads and writes this key, increases `c.readerMap[key]` to `2`.

That will leave `(c.readerMap[key] > 0 && !keyRead) == false` while it should be blocked.

Reply:
In that case it's already blocked since we iterate trough the `rset` first.

Reply:
That was a really nice catch!

Reply:
After some back and forth, decided to go with the simplified version above just with three loops.